railgun/test/integration/knative_test.go

```diff
@@ -0,0 +1,286 @@
+//+build integration_tests
+
+package integration
+
+import (
+	"crypto/tls"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+	"time"
+
+	"bytes"
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+
+	types "k8s.io/apimachinery/pkg/types"
+
+	"github.com/kong/kubernetes-testing-framework/pkg/kind"
+	"github.com/stretchr/testify/assert"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	knativeCrds = "https://github.com/knative/serving/releases/download/v0.13.0/serving-crds.yaml"
+	knativeCore = "https://github.com/knative/serving/releases/download/v0.13.0/serving-core.yaml"
+)
+
+func isKnativeReady(ctx context.Context, cluster kind.Cluster) bool {
+	timeout := 1
+	for timeout < 60 {
+		podList, err := cluster.Client().CoreV1().Pods("knative-serving").List(ctx, metav1.ListOptions{})
+		if err != nil {
+			fmt.Println("failed retrieving knative pods.", err)
+			time.Sleep(1 * time.Second)
+			continue
+		}
+
+		ready := true
+		if len(podList.Items) != 4 {
+			fmt.Println("expected 4 pods, found ", len(podList.Items))
+			time.Sleep(1 * time.Second)
+			continue
+		}
+
+		for _, pod := range podList.Items {
+			if pod.Status.Phase != v1.PodRunning {
+				ready = false
+			}
+		}
+		if ready == true {
+			fmt.Println("All knative pods are up and ready.")
+			fmt.Println("Covering a window that webhook has been configured itself but not ready to receive traffic yet.")
+			time.Sleep(3 * time.Second)
+			return true
+		}
+		if ready == false {
+			time.Sleep(1 * time.Second)
+		}
+	}
+
+	fmt.Println("Failed to bring up knative resources within 60 seconds.")
+	return false
+}
+
+func TestKnativeIngress(t *testing.T) {
+	t.Skip("please use the steps run the test separately. ")
+	ctx := context.Background()
+
+	t.Log("Deploying all resources that are required to run knative")
+	err := deployManifest(knativeCrds, ctx)
+	assert.NoError(t, err)
+	err = deployManifest(knativeCore, ctx)
+	assert.NoError(t, err)
+	knativeReady := isKnativeReady(ctx, cluster)
+	assert.Equal(t, knativeReady, true)
+
+	t.Log("Note down the ip address or public CNAME of kong-proxy service.")
+	proxy, err := retrieveProxyInfo(ctx)
+	if err != nil {
+		t.Fatalf("kong-proxy service ip/public name is not ready.")
+	}
+
+	t.Log("Configure Knative NetworkLayer as Kong")
+	err = configKnativeNetwork(ctx, cluster)
+	assert.NoError(t, err)
+	err = configKnativeDomain(ctx, proxy, cluster)
+	assert.NoError(t, err)
+	// cover webhook sync windows
+	time.Sleep(3 * time.Second)
+
+	t.Log("Install knative service")
+	err = installKnativeSrv(ctx)
+	assert.NoError(t, err)
+
+	t.Log("Test knative service using kong.")
+	srvaccessable := accessKnativeSrv(ctx, proxy)
+	if srvaccessable == false {
+		t.Fatalf("failed to access knative service.")
+	}
+
+	t.Log("clean up test deployments.")
+	deleteManifest(knativeCrds, ctx)
+	deleteManifest(knativeCore, ctx)
+	deleteManifest("helloworldgo.yaml", ctx)
+	time.Sleep(5 * time.Second)
```

Comment:
If this is settle time for teardowns, we could use a `require.Eventually` or `assert.Eventually` here to validate the teardown instead.

Reply:
kindly refer previous response.